### PR TITLE
storage: Added Chunks{Queryable/Querier/SeriesSet/Series/Iteratable. Added generic Merge{SeriesSet/Querier} implementation.

### PR DESCRIPTION
**Pull request by bwplotka**
## Rationales:

In many places (e.g. chunk Remote read, Thanos Receive fetching chunk from TSDB), we operate on encoded chunks not samples.
This means that we unnecessary decode/encode, wasting CPU, time and memory.
This PR adds chunk iterator interfaces and makes the merge code to be reused between both seriesSets

I will make the use of it in following PR inside TSDB itself. For now, fanout implements it and mergers.

All merges now also allow passing series mergers. This opens doors for custom deduplication other than TSDB vertical ones (e.g. offline one we have in Thanos).

## Changes

* Added Chunk versions of alliterating methods. It all starts in Querier/ChunkQuerier. The plan is that
Storage will implement both chunked and samples.
* Added Seek to chunks.Iterator interface for iterating over chunks.
* NewMergeChunkQuerier was added; Both this and NewMergeQuerier are now using generigMergeQuerier to share the code. A generic code was added.
* Improved tests.
* Added some TODO for further simplifications in the next PRs.

cc @krasi-georgiev @brian-brazil @tomwilkie @brancz @pstibrany 

Signed-off-by: Bartlomiej Plotka <bwplotka@gmail.com>

### CODE REVIEW

File: storage/fanout.go [hunk 1]
```diff
@@ -542,15 +631,15 @@ func (c *mergeIterator) Seek(t int64) bool {
 	return len(c.h) > 0
 }
 
-func (c *mergeIterator) At() (t int64, v float64) {
+func (c *chainSampleIterator) At() (t int64, v float64) {
 	if len(c.h) == 0 {
-		panic("mergeIterator.At() called after .Next() returned false.")
+		return math.MaxInt64, 0
```
Comment (brian-brazil) [inline storage/fanout.go]: Panicing here was correct

File: storage/interface.go [hunk 2]
```diff
@@ -51,19 +52,40 @@ type Storage interface {
 }
 
 // A Queryable handles queries against a storage.
+// Use it when you need to have access to all samples without chunk encoding abstraction e.g promQL
```
Comment (brian-brazil) [inline storage/interface.go]: full stop

File: storage/fanout.go [hunk 3]
```diff
@@ -481,59 +554,75 @@ func (c *mergeSeriesSet) Err() error {
...
 	x := old[n-1]
 	*h = old[0 : n-1]
 	return x
 }
 
-type mergeSeries struct {
+// ChainedSeriesMerge returns single series from many same series by chaining samples together.
+// In case of the timestamp overlap, first overlapped sample is kept, rest are dropped.
+// We expect the same labels for each given series.
+// TODO(bwplotka): This has the same logic as tsdb.verticalChainedSeries. Remove this in favor of ChainedSeriesMerge in next PRs.
```
Comment (brian-brazil) [inline storage/fanout.go]: So the general plan is to remove the pluggability of the merge function, and only have vertical merging everywhere?

File: tsdb/chunks/chunks.go [hunk 4]
```diff
@@ -67,6 +67,20 @@ type Meta struct {
 	MinTime, MaxTime int64
 }
 
+// Iterator iterates over the chunk of a time series.
+type Iterator interface {
+	//// Seek advances the iterator forward to the given timestamp.
+	//// It advances to the chunk with min time at t or first chunk with min time after t.
+	//Seek(t int64) bool
```
Comment (brian-brazil) [inline tsdb/chunks/chunks.go]: Is this meant to be commented out?

File: storage/series.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+type chunkSetToSeriesSet struct {
+	ChunkSeriesSet
+
+	chkIterErr       error
+	sameSeriesChunks []Series
+	bufIterator      chunkenc.Iterator
+}
+
+// NewSeriesSetFromChunkSeriesSet use ChunkSeriesSet to provider sample SeriesSet.
```
Comment (brian-brazil) [inline storage/series.go]: This isn't clear

File: tsdb/chunks/chunks.go [hunk 6]
```diff
@@ -67,6 +67,20 @@ type Meta struct {
 	MinTime, MaxTime int64
 }
 
+// Iterator iterates over the chunk of a time series.
+type Iterator interface {
```
Comment (brian-brazil) [inline tsdb/chunks/chunks.go]: This isn't used anywhere in this module, is this the right place for it?

File: storage/fanout.go [hunk 7]
```diff
@@ -590,22 +680,164 @@ func (c *mergeIterator) Err() error {
...
+	*h = old[0 : n-1]
+	return x
+}
+
+// VerticalChunkMergeFunc represents a function that merges multiple time overlapping chunks.
+// Passed chunks:
+// * have to be sorted by MinTime.
+// * have to be part of exactly the same timeseries.
+// * have to be populated.
+// Merge can process can result in more than once chunk.
```
Comment (brian-brazil) [inline storage/fanout.go]: Extra can

File: storage/generic.go [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+	if q.q != nil {
+		s, w, err := q.q.Select(sortSeries, hints, matchers...)
+		return &genericSeriesSetAdapter{s}, w, err
+	}
+	s, w, err := q.cq.Select(sortSeries, hints, matchers...)
+	return &genericChunkSeriesSetAdapter{s}, w, err
+}
+
+func newGenericQuerierFrom(q Querier) genericQuerier {
+	if q == nil {
```
Comment (brian-brazil) [inline storage/generic.go]: When will this be nil?

File: storage/generic.go [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+	}
+	return &genericQuerierAdapter{baseQuerier: cq, cq: cq}
+}
+
+type querierAdapter struct {
+	genericQuerier
+}
+
+type seriesSetAdapter struct {
+	genericSeriesSet
```
Comment (brian-brazil) [inline storage/generic.go]: It's a bit odd for this to be depending on something defined in fanout.go

File: storage/generic.go (hunk 8 above)
Comment (bwplotka) [reply storage/generic.go]: I was thinking about this case: https://github.com/prometheus/prometheus/blob/4afc2f3e6be7c3f605a5f091c45de74cadd8cada/storage/remote/storage.go#L148 

but full nil probably will never happen - also will result in panic in the caller of `newGenericQuerierFrom` anyway.

Will remove (: thanks.

**Comment by bwplotka** [comment]
You have @brian-brazil good question so I will put it in more visible place here:

> So the general plan is to remove the pluggability of the merge function, and only have vertical merging everywhere?

Yes. Because I don't get why we had two different iterators (vertical and normal). Planned vertical one does not add any overhead for non-overlapping samples.

In fact, I also plan to use **the same** chained merge (which does vertical merge) for both querier AND compaction.. so we will have just ONE merge logic, not four like we used to have. Because with four, once I would add chunk iteration we would have 8.... (:

It's also pluggable in terms of what `merge` logic you want. Especially for chunks, you have a function that does it in old fashion (chain all chunks samples together) and for Thanos we will be able to provide more advanced one for offline deduplication as part of https://github.com/thanos-io/thanos/issues/1014 :hugs:

File: storage/fanout.go (hunk 3 above)
Comment (bwplotka) [reply storage/fanout.go]: Correct. Commented here: https://github.com/prometheus/prometheus/pull/7005#issuecomment-602475951

File: tsdb/chunks/chunks.go (hunk 4 above)
Comment (bwplotka) [reply tsdb/chunks/chunks.go]: Sorry! I just did not know if we really want Seek... I don't think we do. Removing.

File: storage/series.go (hunk 5 above)
Comment (bwplotka) [reply storage/series.go]: fixed.

File: tsdb/chunks/chunks.go (hunk 6 above)
Comment (bwplotka) [reply tsdb/chunks/chunks.go]: Hm, yes as it is a nice place that everyone depends on anyway for chunks.

Comment (bwplotka) [reply tsdb/chunks/chunks.go]: This is really the same as `chunkenc.Iterator`

File: storage/generic.go (hunk 9 above)
Comment (bwplotka) [reply storage/generic.go]: Moved here.

**Comment by bwplotka** [comment]
Addressed all @brian-brazil , thanks for quick review! Should be ready for another round :muscle:

File: storage/series.go [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+	"github.com/prometheus/prometheus/tsdb/tsdbutil"
+)
+
+// listSeriesIterator allows to iterate over provided samples. Does not handle overlaps.
+type listSeriesIterator struct {
+	samples []tsdbutil.Sample
+	idx     int
+}
+
+func NewListSeriesIterator(samples []tsdbutil.Sample) chunkenc.Iterator {
```
Comment (krasi-georgiev) [inline storage/series.go]: exported and no comment

File: storage/series.go [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+func (c *chunkSetToSeriesSet) Err() error {
+	if c.chkIterErr != nil {
+		return c.chkIterErr
+	}
+	return c.ChunkSeriesSet.Err()
+}
+
+var _ Series = &chunkToSeries{}
```
Comment (krasi-georgiev) [inline storage/series.go]: Why is this needed?

```suggestion
```

File: storage/series_test.go [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+import (
+	"math"
+
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/tsdb/chunkenc"
+	"github.com/prometheus/prometheus/tsdb/chunks"
+	"github.com/prometheus/prometheus/tsdb/tsdbutil"
+)
+
+var _ Series = &MockSeries{}
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/series_test.go]: To ensure mock is actually implementing interface we want.

File: storage/generic.go [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+	q  Querier
+	cq ChunkQuerier
+}
+
+type genericSeriesSetAdapter struct {
+	SeriesSet
+}
+
+func (a *genericSeriesSetAdapter) At() SeriesLabels {
+	return a.SeriesSet.At().(SeriesLabels)
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/generic.go]: You have to cast it otherwise it will not work.

[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/generic.go]: (passing interface to another interface has to be converted explicitly)

File: storage/generic.go [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+type genericChunkSeriesSetAdapter struct {
+	ChunkSeriesSet
+}
+
+func (a *genericChunkSeriesSetAdapter) At() SeriesLabels {
+	return a.ChunkSeriesSet.At().(SeriesLabels)
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/generic.go]: that will not complile

File: storage/generic.go [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+
+func newGenericQuerierFrom(q Querier) genericQuerier {
+	return &genericQuerierAdapter{baseQuerier: q, q: q}
+}
+
+func newGenericQuerierFromChunk(cq ChunkQuerier) genericQuerier {
+	return &genericQuerierAdapter{baseQuerier: cq, cq: cq}
+}
+
+type querierAdapter struct {
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/generic.go]: What? (:

File: storage/fanout.go [hunk 16]
```diff
@@ -181,66 +183,94 @@ func (f *fanoutAppender) Rollback() (err error) {
...
-// mergeQuerier implements Querier.
-type mergeQuerier struct {
-	primaryQuerier Querier
-	queriers       []Querier
+type mergeGenericQuerier struct {
+	mergeFunc genericSeriesMergeFunc
 
-	failedQueriers map[Querier]struct{}
-	setQuerierMap  map[SeriesSet]Querier
+	primaryQuerier genericQuerier
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/fanout.go]: I have no idea why I can only guess so will leave that to another PRs (:

File: storage/fanout.go [hunk 17]
```diff
@@ -590,22 +666,164 @@ func (c *mergeIterator) Err() error {
...
+type VerticalChunksMergeFunc func(chks ...chunks.Meta) chunks.Iterator
+
+type verticalChunkSeriesMerger struct {
+	verticalChunksMerger VerticalChunksMergeFunc
+
+	labels labels.Labels
+	series []ChunkSeries
+}
+
+// NewVerticalChunkSeriesMerger returns VerticalChunkSeriesMerger that merges the same chunk series into one chunk
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/fanout.go]: Will fix :+1:

File: storage/fanout.go [hunk 18]
```diff
@@ -590,22 +666,164 @@ func (c *mergeIterator) Err() error {
...
+// chainChunkIterator is responsible to chain chunks from different iterators of same time series.
+// If they are time overlapping overlappedChunksMerger will be used.
+type chainChunkIterator struct {
+	overlappedChunksMerger VerticalChunksMergeFunc
+
+	iterators []chunks.Iterator
+	h         chunkIteratorHeap
+}
+
+func (c *chainChunkIterator) At() chunks.Meta {
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/fanout.go]: true, will fix

File: storage/series.go (hunk 11 above)
Comment (bwplotka) [reply storage/series.go]: commented above.

File: storage/series_test.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply storage/series_test.go]: git it, maybe a comment?

File: storage/fanout_test.go [hunk 19]
```diff
@@ -52,223 +54,525 @@ func TestMergeTwoStringSlices(t *testing.T) {
 	}
 }
 
-func TestMergeSeriesSet(t *testing.T) {
+func TestMergeQuerierWithChainMerger(t *testing.T) {
 	for _, tc := range []struct {
-		input    []SeriesSet
+		name          string
+		querierSeries [][]Series
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/fanout_test.go]: yes but in this form.... those are still queriers' series ):

[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/fanout_test.go]: so I would vote to keep it

File: storage/fanout_test.go [hunk 20]
```diff
@@ -52,223 +54,525 @@ func TestMergeTwoStringSlices(t *testing.T) {
...
+				NewListSeries(labels.FromStrings("bar", "baz"), []tsdbutil.Sample{sample{2, 21}, sample{3, 31}, sample{5, 5}, sample{6, 6}}),
+				NewListSeries(labels.FromStrings("foo", "bar"), []tsdbutil.Sample{sample{0, 0}, sample{1, 1}, sample{2, 2}}),
+			}, {
+				NewListSeries(labels.FromStrings("bar", "baz"), []tsdbutil.Sample{sample{1, 1}, sample{2, 22}, sample{3, 32}}),
+				NewListSeries(labels.FromStrings("foo", "bar"), []tsdbutil.Sample{sample{3, 3}, sample{4, 4}}),
+			}, {}},
+			expected: NewMockSeriesSet(
+				NewListSeries(
+					labels.FromStrings("bar", "baz"),
+					[]tsdbutil.Sample{sample{1, 1}, sample{2, 21}, sample{3, 31}, sample{5, 5}, sample{6, 6}},
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/fanout_test.go]: yes

File: storage/fanout_test.go [hunk 21]
```diff
@@ -52,223 +54,525 @@ func TestMergeTwoStringSlices(t *testing.T) {
...
 		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			var qs []Querier
+			for _, in := range tc.querierSeries {
+				qs = append(qs, &mockQuerier{toReturn: in})
+			}
+			qs = append(qs, tc.extraQueriers...)
+
+			merged, _, _ := NewMergeQuerier(qs[0], qs, ChainedSeriesMerge).Select(false, nil)
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/fanout_test.go]: Nope, but this is confusing agree. This is how it was designed from start.

File: storage/series_test.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/series_test.go]: Hm I would say let's not comment (:  I won't be able to comment 10000 other places in this codebase and in any other Go project, it's really known usually pattern.

File: storage/fanout_test.go [hunk 22]
```diff
@@ -52,223 +54,525 @@ func TestMergeTwoStringSlices(t *testing.T) {
...
+				testutil.Equals(t, expErr, actErr)
+				testutil.Equals(t, expSmpl, actSmpl)
+			}
+			testutil.Ok(t, merged.Err())
+			testutil.Assert(t, !tc.expected.Next(), "Expected Next() to be false")
+		})
+	}
+}
+
+func TestMergeChunkQuerierWithNoVerticalChunkSeriesMerger(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/fanout_test.go]: Because there is no vertical merger implementation in this package. This is tested outside.

File: storage/generic.go [hunk 23]
```diff
@@ -0,0 +1,133 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storage
+
+// This file holds boilerplate adapters for generic MergeSeriesSet and MergeQuerier functions, so we can have one optimized
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/generic.go]: :+1:

File: storage/generic.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/generic.go]: ah got it

File: storage/interface.go [hunk 24]
```diff
@@ -149,19 +171,43 @@ func (s errSeriesSet) Next() bool { return false }
...
+	Err() error
+}
+
+// ChunkSeries exposes a single time series and allows iterating over chunks.
+type ChunkSeries interface {
+	SeriesLabels
+	ChunkIteratable
+}
+
+// SeriesLabels represents the item that have labels e.g. time series.
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/interface.go]: I can name it `Labels` then? (:

File: storage/fanout_test.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply storage/fanout_test.go]: for the same series there are `sample{3, 31}`  and `sample{3, 32}` why is `sample{3, 31}` selected?

File: storage/fanout_test.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply storage/fanout_test.go]: sorry which question did you answer here? the first one or the second?

File: storage/fanout.go [hunk 25]
```diff
@@ -68,24 +70,24 @@ func (f *fanout) StartTime() (int64, error) {
...
 	}
 	queriers = append(queriers, primaryQuerier)
 
-	// Add secondary queriers
+	// Add secondary queriers.
 	for _, storage := range f.secondaries {
 		querier, err := storage.Querier(ctx, mint, maxt)
 		if err != nil {
-			NewMergeQuerier(primaryQuerier, queriers).Close()
+			NewMergeQuerier(primaryQuerier, queriers, VerticalSeriesMergeFunc(ChainedSeriesMerge)).Close()
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/fanout.go]: fair

File: storage/fanout.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/fanout.go]: added TODO!

File: storage/fanout.go [hunk 26]
```diff
@@ -373,39 +403,70 @@ func (q *mergeQuerier) LabelNames() ([]string, Warnings, error) {
...
+	mergeFunc     genericSeriesMergeFunc
+
+	heap genericSeriesSetHeap
+	sets []genericSeriesSet
+
+	currentSets []genericSeriesSet
+	querier     *mergeGenericQuerier
+}
+
+// Merge returns merged series implementation that merges series with same labels together.
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/fanout.go]: nice catch :+1:

File: storage/series_test.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply storage/series_test.go]: I didn't know it was a pattern. But why do you even need this check? I mean the test compilation will fail anyway if the `MockSeries` don't implement the `Series` interface as you expect a slice of `[]Series` there

File: storage/generic.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply storage/generic.go]: I tried it before suggesting and it works :)

File: storage/generic.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply storage/generic.go]: it does :)

File: storage/interface.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply storage/interface.go]: When reading all the code I think the name  `GenericSeries` fits best and it is a bit more consistent with the naming of the other interfaces, but again this is a probably how I see the code and someone else will see read it differently so feel free to keep as is if you prefer.

File: storage/fanout.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply storage/fanout.go]: I followed the code and that is the reason. Failing primary queried returns an error and falling secondary queries return just warning.

File: storage/fanout.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/fanout.go]: wait, you did not read the sentence till the end (:

File: storage/fanout_test.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/fanout_test.go]: Second (:

File: storage/series_test.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/series_test.go]: fair point

File: storage/generic.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/generic.go]: That's funny, so you need to cast explictly if the interfaces is smaller but not if larger (: That's good to know! (:

File: storage/fanout_test.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/fanout_test.go]: > hm I didn't know that passing nil for the matchers returns all series. Maybe add this in the Select interface comment

That's not correct. No matchers will give you nothing. It's just because everything is mocked, this works.

File: storage/fanout_test.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/fanout_test.go]: I don't think we should really note this behaviour - I am not sure if it's deterministic really. Do we care?

File: storage/generic.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/generic.go]: lolol

File: storage/fanout.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/fanout.go]: Ah, ok  will note

File: storage/fanout.go [hunk 27]
```diff
@@ -373,39 +407,70 @@ func (q *mergeQuerier) LabelNames() ([]string, Warnings, error) {
...
 }
 
-// NewMergeSeriesSet returns a new series set that merges (deduplicates)
-// series returned by the input series sets when iterating.
-// Each input series set must return its series in labels order, otherwise
+// VerticalSeriesMergeFunc returns merged series implementation that merges series with same labels together.
+// It has to handle time-overlapped series as well.
+type VerticalSeriesMergeFunc func(...Series) Series
+
+// VerticalSeriesMergeFunc returns merged chunk series implementation that merges series with same labels together.
```
Comment (krasi-georgiev) [inline storage/fanout.go]:
```suggestion
// VerticalChunkSeriesMergerFunc returns merged chunk series implementation that merges series with same labels together.
```

File: storage/fanout_test.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply storage/fanout_test.go]: you have more context here so If you don't think it matters than all good.

File: storage/fanout.go [hunk 28]
```diff
@@ -373,39 +407,70 @@ func (q *mergeQuerier) LabelNames() ([]string, Warnings, error) {
...
+	}
+	return &chunkSeriesSetAdapter{newGenericMergeSeriesSet(genericSets, nil, (&chunkSeriesMergerAdapter{VerticalChunkSeriesMergerFunc: merger}).Merge)}
+}
+
+// newGenericMergeSeriesSet returns a new genericSeriesSet that merges (and deduplicates)
+// series returned by the chkQuerierSeries series sets when iterating.
+// Each chkQuerierSeries series set must return its series in labels order, otherwise
 // merged series set will be incorrect.
-func NewMergeSeriesSet(sets []SeriesSet, querier *mergeQuerier) SeriesSet {
+// Argument 'querier' is optional and can be nil. Pass it to query if you have failing sets.
```
Comment (krasi-georgiev) [inline storage/fanout.go]: Hm, I still don't understand this last sentence.

File: storage/series.go (hunk 11 above)
Comment (krasi-georgiev) [reply storage/series.go]: I double checked and this guarantees that `chunkToSeries` implements `Series`

https://github.com/prometheus/prometheus/pull/7005/files#diff-dde62d868fca6daee199a54f46f8d8c4R104

File: storage/fanout.go (hunk 27 above)
Comment (bwplotka) [reply storage/fanout.go]: :man_facepalming:

**Comment by krasi-georgiev** [comment]
LGTM

**Comment by bwplotka** [comment]
:tada: Thanks all for quick review!
